File: Design-Patterns-Using-Kotlin/Interpreter/InterpreterPattern.kt
// Abstract Expression
interface AbstractExpression {
    fun interpret()
}

// Nonterminal Expression
class NonterminalExpression(private val expression: AbstractExpression) : AbstractExpression {
    override fun interpret() {
        println("NonTerminalExpression:interpret")
        expression.interpret()
    }
}

// Terminal Expression
class TerminalExpression : AbstractExpression {
    override fun interpret() {
        println("TerminalExpression:interpret")
    }
}

// Client Code
fun main() {
    val tree: AbstractExpression = NonterminalExpression(TerminalExpression())
    tree.interpret()
}

/*
NonTerminalExpression:interpret
TerminalExpression:interpret 
*/

File: Design-Patterns-Using-Ruby/Interpreter/InterpreterPattern.rb
# Abstract Expression
class AbstractExpression
    def interpret
        raise NotImplementedError, 'Subclasses must implement the interpret method'
    end
end

# Nonterminal Expression
class NonterminalExpression < AbstractExpression
    def initialize(expression)
        @expression = expression
    end

    def interpret
        puts 'NonTerminalExpression:interpret'
        @expression.interpret
    end
end

# Terminal Expression
class TerminalExpression < AbstractExpression
    def interpret
        puts 'TerminalExpression:interpret'
    end
end

# Client Code
tree = NonterminalExpression.new(TerminalExpression.new)
tree.interpret

=begin 
NonTerminalExpression:interpret
TerminalExpression:interpret
=end